### PR TITLE
oshmem: add C++ wrapper compilers

### PR DESCRIPTION
**Pull request by ggouaillardet**
though there are no C++ bindings for oshmem, we need C++ wrappers
since a C compiler might not be able to compile a C++ source.
the C++ wrappers are :
- shmemc++ / oshc++
- shmemcxx / oshcxx
- shmemCC / oshCC (on case sensitive filesystems)

Thanks Bert Wesarg for bringing this to our attention

### CODE REVIEW

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by ggouaillardet** [comment]
@jladd-mlnx can you please review this ? then will PR vs the ```v2``` branches

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by jsquyres** [comment]
@ggouaillardet This is not a bug; it's a feature enhancement.  As such, I moved the milestone to v2.1.0.

**Comment by ggouaillardet** [comment]
@jsquyres the initial report states our README mentions ```oshCC``` 
I am also fine to consider the bug is the README, and not the lack of ```oshCC``` and friends

**Comment by jsquyres** [comment]
Oh, I missed the mention of oshCC -- gotcha.

Yes, I think this is a README bug (in terms of 2.0.x).  It's a new feature, since we've never had it before, so the new oshCC wrapper itself should be headed for v2.1.0.

**Comment by ggouaillardet** [comment]
:bot:mellanox:retest

**Comment by jladd-mlnx** [comment]
bot:mellanox:retest

**Comment by jladd-mlnx** [comment]
@ggouaillardet This looks good. I'm not sure why jenkins is failing - I'll need to dig some more. I can manually run the failing launch command.

File: oshmem/tools/wrappers/Makefile.am [hunk 1]
```diff
@@ -12,49 +12,95 @@
...
 	$(DESTDIR)$(pkgdatadir)/shmemfort-wrapper-data.txt \
 	$(DESTDIR)$(pkgdatadir)/oshfort-wrapper-data.txt
 
 
+if CASE_SENSITIVE_FS
+man_MANS += oshCC.1 shmemCC.1
+
+install-exec-hook: install-exec-hook-always
+	(cd $(DESTDIR)$(bindir); rm -f shmemCC$(EXEEXT); $(LN_S) mpiCC$(EXEEXT) shmemCC$(EXEEXT))
+	(cd $(DESTDIR)$(bindir); rm -f oshCC$(EXEEXT); $(LN_S) mpiCC$(EXEEXT) oshCC$(EXEEXT))
```
Comment (jsquyres) [inline oshmem/tools/wrappers/Makefile.am]: These should link to `opal_wrapper`, not `mpiCC`.

File: oshmem/tools/wrappers/Makefile.am [hunk 2]
```diff
@@ -12,49 +12,95 @@
...
-	(cd $(DESTDIR)$(bindir); rm -f shmemrun$(EXEEXT); $(LN_S) mpirun$(EXEEXT) shmemrun$(EXEEXT))
-	(cd $(DESTDIR)$(bindir); rm -f oshrun$(EXEEXT); $(LN_S) mpirun$(EXEEXT) oshrun$(EXEEXT))
-	(cd $(DESTDIR)$(bindir); rm -f shmemcc$(EXEEXT); $(LN_S) mpicc$(EXEEXT) shmemcc$(EXEEXT))
-	(cd $(DESTDIR)$(bindir); rm -f oshcc$(EXEEXT); $(LN_S) mpicc$(EXEEXT) oshcc$(EXEEXT))
-	(cd $(DESTDIR)$(bindir); rm -f shmemfort$(EXEEXT); $(LN_S) mpifort$(EXEEXT) shmemfort$(EXEEXT))
-	(cd $(DESTDIR)$(bindir); rm -f oshfort$(EXEEXT); $(LN_S) mpifort$(EXEEXT) oshfort$(EXEEXT))
-
-install-data-hook:
+	(cd $(DESTDIR)$(bindir); rm -f shmemrun$(EXEEXT); $(LN_S) opal_wrapper$(EXEEXT) shmemrun$(EXEEXT))
+	(cd $(DESTDIR)$(bindir); rm -f oshrun$(EXEEXT); $(LN_S) opal_wrapper$(EXEEXT) oshrun$(EXEEXT))
```
Comment (jsquyres) [inline oshmem/tools/wrappers/Makefile.am]: Shouldn't these first two be linked against `mpirun` (or at least `orterun`)?

**Comment by ggouaillardet** [comment]
@jsquyres good catches, i made the requested changes
i also pushed one more commit to cleanup ```oshmem_info``` output
(no C++ binding, and only one flavor of Fortran (```include 'shmem.fh'```)

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by ibm-ompi** [comment]
Build Failed with GNU compiler! Please review the log, and get in touch if you have questions.

 Gist: https://gist.github.com/01503faf56097fbbc531c18bd107ef42

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by ibm-ompi** [comment]
Build Failed with PGI compiler! Please review the log, and get in touch if you have questions.

 Gist: https://gist.github.com/b5243b56c89c41256c8f0cfd45e61bc2

**Comment by ibm-ompi** [comment]
Build Failed with XL compiler! Please review the log, and get in touch if you have questions.

 Gist: https://gist.github.com/1d3cad2487eb6ac9ce50d3da89d224c2

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by jsquyres** [comment]
@ggouaillardet Won't your oshmem_info changes potentially break backwards compatibility?  I.e., if anyone is grepping for that output in oshmem_info -- regardless of whether it's meaningful for OSHMEM or not -- that will break, no?

**Comment by ggouaillardet** [comment]
@jsquyres it will obviously impact backward compatibility.
that is why i split the PR into two commits, so only the first one will land into v2.x

**Comment by jsquyres** [comment]
@ggouaillardet Have you had a chance to split this into what can go into v2.x and what can't yet, perchance?

**Comment by ggouaillardet** [comment]
- open-mpi/ompi@810100f7375714b554100b9a1343faad75501985 is good for v2.x
strictly speaking, this is an enhancement, so milestone could be v2.1.0 or v2.2.0 if feature is already frozen
- open-mpi/ompi@835f855ecda6405238ee1ee048a8a5daeb758068 breaks ```oshmem_info``` backward compatibility, so it will be for v3.0.0

i wrote "i split" and meant to be read as simple past, so i __did__ split the PR into two commits a couple weeks ago

**Comment by jsquyres** [comment]
Oooh... this is a PR on master, not v2.x.  Got it.  Yes, this makes sense now.

**Comment by jsquyres** [comment]
@ggouaillardet In the F2F discussion in Jan 2017, we'd love to see this move forward.  Any chance of rebasing and finishing this PR?

File: examples/hello_oshmem_cxx.cc [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ *  * Copyright (c) 2014      Mellanox Technologies, Inc.
+ *   *                         All rights reserved.
+ *    * Copyright (c) 2015 Cisco Systems, Inc.  All rights reserved.
+ *     * $COPYRIGHT$
+ *      *
+ *       * Additional copyrights may follow
+ *        *
+ *         * $HEADER$
+ *          */
+
```
Comment (bertwesarg) [inline examples/hello_oshmem_cxx.cc]: Is this comment style on purpose?

**Comment by ggouaillardet** [comment]
@bertwesarg i fixed that, thanks for the heads up !

**Comment by jsquyres** [comment]
@ggouaillardet Sorry I missed this one -- you signed it off long ago; it's now good to go.  I think we'll have to skip it for v2.1.x, though -- it can be slated for v3.0.0 (which means you don't need to do anything other than merge this to master in the near future -- v3.x may be branched from master "soon").

**Comment by ibm-ompi** [comment]
The IBM CI (GNU Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/7d4702e07e59e3bcd661f498a514bf1d

**Comment by jjhursey** [comment]
Weird. I don't know why the IBM CI triggered on this PR. It's been out here for a long time, and has not recent traffic.

**Pull request merged**